pkg/controllers/infra/cluster_controller.go

```diff
@@ -0,0 +1,229 @@
+}
+
+func (r *ClusterController) deleteClusterCredentialsIfNeeded(ctx context.Context, infraCluster *infrav1.Cluster) error {
+	secretList := &corev1.SecretList{}
+	if err := r.List(ctx, secretList, client.InNamespace(system.GetManagerNamespace()), client.MatchingLabels{
+		clusterNameLabel: infraCluster.Name,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to list Secrets in namespace %s", infraCluster.Namespace)
+	}
+
+	for _, secret := range secretList.Items {
+		if err := r.Delete(ctx, &secret); err != nil {
+			if apiserrors.IsNotFound(err) {
+				continue
+			}
+			return errors.Wrapf(err, "failed to delete Secret %s/%s", secret.Namespace, secret.Name)
+		}
+	}
+
+	identityLiest := &capav1.AWSClusterStaticIdentityList{}
+	if err := r.List(ctx, identityLiest, client.InNamespace(""), client.MatchingLabels{
+		clusterNameLabel: infraCluster.Name,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to list Secrets in namespace %s", infraCluster.Namespace)
+	}
+	for _, identity := range identityLiest.Items {
+		if err := r.Delete(ctx, &identity); err != nil {
+			if apiserrors.IsNotFound(err) {
+				continue
+			}
+			return errors.Wrapf(err, "failed to delete AWSClusterStaticIdentity %s", identity.Name)
+		}
+	}
+
+	return nil
+}
+
+func (r *ClusterController) deleteClusterResourceSet(ctx context.Context, infraCluster *infrav1.Cluster) error {
+	clusterResourceSetList := &addonsv1.ClusterResourceSetList{}
+	if err := r.List(ctx, clusterResourceSetList, client.InNamespace(infraCluster.Namespace), client.MatchingLabels{
+		clusterNameLabel: infraCluster.Name,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to list ClusterResourceSet in namespace %s", infraCluster.Namespace)
+	}
+
+	for _, clusterResourceSet := range clusterResourceSetList.Items {
+		if err := r.Delete(ctx, &clusterResourceSet); err != nil {
+			if apiserrors.IsNotFound(err) {
+				continue
+			}
+			return errors.Wrapf(err, "failed to delete ClusterResourceSet %s/%s", clusterResourceSet.Namespace, clusterResourceSet.Name)
+		}
+	}
+
+	return nil
+}
+
+func (r *ClusterController) deleteCAPIClusterIfNeeded(ctx context.Context, infraCluster *infrav1.Cluster) error {
+	cluster := &capiv1.Cluster{}
+	nn := types.NamespacedName{
+		Namespace: infraCluster.Namespace,
+		Name:      infraCluster.Name,
+	}
+	if err := r.Get(ctx, nn, cluster); err != nil {
+		if apiserrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to get CAPI Cluster %s/%s", infraCluster.Namespace, infraCluster.Name)
+	}
+
+	if err := r.Delete(ctx, cluster); err != nil {
+		if apiserrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to delete CAPI Cluster %s/%s", infraCluster.Namespace, infraCluster.Name)
+	}
+
+	return nil
+}
+
+func (r *ClusterController) reconcile(ctx context.Context, infraCluster *infrav1.Cluster) (ctrl.Result, error) {
+	switch infraCluster.Spec.InfraType {
+	case infrav1.AWSClusterInfraType:
+		panic("not implemented")
```

Comment:
?? This is the target work, right

Reply:
mark this as TODO, will implement it later.